src/libs/import.c

```diff
@@ -480,166 +508,581 @@ static void _lib_import_update_preview(GtkFileChooser *file_chooser, gpointer da
+                                          DT_IMPORT_THUMB, pixbuf, -1);
+
+  if(pixbuf) g_object_ref(pixbuf);
   g_free(filename);
+  g_free(fullname);
+}
 
-  gtk_file_chooser_set_preview_widget_active(file_chooser, have_preview);
+static gboolean _thumb_toggled(GtkWidget *view, GdkEventButton *event, dt_lib_module_t *self)
+{
+  dt_lib_import_t *d = (dt_lib_import_t *)self->data;
+  if((event->type == GDK_BUTTON_PRESS && event->button == 1))
+  {
+    GtkTreePath *path = NULL;
+    GtkTreeViewColumn *column = NULL;
+    // Get tree path for row that was clicked
+    if(gtk_tree_view_get_path_at_pos(GTK_TREE_VIEW(view), (gint)event->x, (gint)event->y,
+                                     &path, &column, NULL, NULL))
+    {
+      if(column == d->from.pixcol)
+      {
+        GtkTreeIter iter;
+        GtkTreeModel *model = GTK_TREE_MODEL(d->from.store);
+        gtk_tree_model_get_iter(model, &iter, path);
+        gboolean thumb_sel;
+        gtk_tree_model_get(model, &iter, DT_IMPORT_SEL_THUMB, &thumb_sel, -1);
+        _thumb_set_in_listview(model, &iter, !thumb_sel, self);
+        return TRUE;
+      }
+    }
+    gtk_tree_path_free(path);
+  }
+  return FALSE;
 }
 
-static void _lib_import_single_image_callback(GtkWidget *widget, dt_lib_import_t* d)
+static gboolean _thumb_set(gpointer user_data)
 {
-  GtkWidget *win = dt_ui_main_window(darktable.gui->ui);
-  GtkWidget *filechooser = gtk_file_chooser_dialog_new(
-      _("import image"), GTK_WINDOW(win), GTK_FILE_CHOOSER_ACTION_OPEN, _("_cancel"), GTK_RESPONSE_CANCEL,
-      _("_open"), GTK_RESPONSE_ACCEPT, (char *)NULL);
-#ifdef GDK_WINDOWING_QUARTZ
-  dt_osx_disallow_fullscreen(filechooser);
-#endif
+  dt_lib_module_t *self = (dt_lib_module_t *)user_data;
+  dt_lib_import_t *d = (dt_lib_import_t *)self->data;
+
+  if(d->from.event)
+  {
+    GtkTreeModel *model = GTK_TREE_MODEL(d->from.store);
+    gboolean thumb_sel;
+    gtk_tree_model_get(model, &d->from.iter, DT_IMPORT_SEL_THUMB, &thumb_sel, -1);
+    if(!thumb_sel)
+      _thumb_set_in_listview(model, &d->from.iter, TRUE, self);
+    if(d->from.event && gtk_tree_model_iter_next(model, &d->from.iter))
+      return TRUE;
+  }
+  d->from.event = 0;
+  return FALSE;
+}
 
-  gtk_file_chooser_set_select_multiple(GTK_FILE_CHOOSER(filechooser), TRUE);
+static void _all_thumb_toggled(GtkTreeViewColumn *column, dt_lib_module_t *self)
+{
+  GtkWidget *toggle = gtk_tree_view_column_get_widget(column);
+  gboolean thumb_sel = !gtk_toggle_button_get_active(GTK_TOGGLE_BUTTON(toggle));
+  gtk_toggle_button_set_active(GTK_TOGGLE_BUTTON(toggle), thumb_sel);
 
-  char *last_directory = dt_conf_get_string("ui_last/import_last_directory");
-  if(last_directory != NULL)
+  dt_lib_import_t *d = (dt_lib_import_t *)self->data;
+  if(!thumb_sel)
+  {
+    // remove the thumbnails
+    d->from.event = 0;
+    GtkTreeModel *model = GTK_TREE_MODEL(d->from.store);
+    GtkTreeIter iter;
+    gboolean valid = gtk_tree_model_get_iter_first(model, &iter);
```

Comment:
given recent work from @ralfbrown  to replace while -> for I would rewrite this:

```
for(gboolean valid = gtk_tree_model_get_iter_first(model, &iter); valid; gtk_tree_model_iter_next(model, &iter))
{
   _thumb_set_in_listview(model, &iter, FALSE, self);
}
```

Reply:
That would need to be
`for(gboolean valid = gtk_tree_model_get_iter_first(model, &iter); valid; valid = gtk_tree_model_iter_next(model, &iter))`
or it will loop forever.

Reply:
Yes of course! Thanks for catching that.

Reply:
Done